pyro/distributions/transforms/spline.py

```diff
@@ -0,0 +1,352 @@
+
+    where :math:`\\alpha^{(k)}` and :math:`\\beta^{(k)}` are two polynomials of order :math:`d`. For :math:`d=1`,
+    we say that the spline is linear, and for :math:`d=2`, quadratic. The spline is constructed on the specified
+    bounding box, :math:`[-K,K]\\times[-K,K]`, with the identity function used elsewhere.
+
+    Rational splines offer an excellent combination of functional flexibility whilst maintaining a numerically
+    stable inverse that is of the same computational and space complexities as the forward operation. This
+    element-wise transform permits the accurate represention of complex univariate distributions.
+
+    Example usage:
+
+    >>> base_dist = dist.Normal(torch.zeros(10), torch.ones(10))
+    >>> transform = Spline(10, count_bins=4, bound=3.)
+    >>> pyro.module("my_transform", transform)  # doctest: +SKIP
+    >>> flow_dist = dist.TransformedDistribution(base_dist, [transform])
+    >>> flow_dist.sample()  # doctest: +SKIP
+        tensor([-0.4071, -0.5030,  0.7924, -0.2366, -0.2387, -0.1417,  0.0868,
+                0.1389, -0.4629,  0.0986])
+
+    :param input_dim: Dimension of the input vector. Despite operating element-wise, this is required so we know
+        how many parameters to store.
+    :type input_dim: int
+    :param count_bins: The number of segments comprising the spline.
+    :type count_bins: int
+    :param bound: The quantity :math:`K` determining the bounding box, :math:`[-K,K]\\times[-K,K]`, of the spline.
+    :type bound: float
+    :param order: One of ['linear', 'quadratic'] specifying the order of the spline.
+    :type order: string
+
+    References:
+
+    Conor Durkan, Artur Bekasov, Iain Murray, George Papamakarios. Neural Spline Flows. NeurIPS 2019.
+
+    Hadi M. Dolatabadi, Sarah Erfani, Christopher Leckie. Invertible Generative Modeling using Linear
+    Rational Splines. AISTATS 2020.
+
+    """
+
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
+    event_dim = 0
+
+    def __init__(self, input_dim, count_bins=8, bound=3., order='linear'):
+        super(Spline, self).__init__(cache_size=1)
+
+        self.layer = SplineLayer(input_dim, count_bins=count_bins, bound=bound, order=order)
+        self._cache_log_detJ = None
+
+    def _call(self, x):
+        y, log_detJ = self.layer(x, jacobian=True)
+        self._cache_log_detJ = log_detJ
+        return y
+
+    def _inverse(self, y):
+        """
+        :param y: the output of the bijection
+        :type y: torch.Tensor
+
+        Inverts y => x. Uses a previously cached inverse if available, otherwise performs the inversion afresh.
+        """
+        x, log_detJ = self.layer(y, jacobian=True, inverse=True)
+        self._cache_log_detJ = -log_detJ
+        return x
+
+    def log_abs_det_jacobian(self, x, y):
+        """
+        Calculates the elementwise determinant of the log jacobian
+        """
+        return self._cache_log_detJ
```

Comment:
I think it's safer to use a caching mechanism similar to the `Transform._cache` and check that inputs match the cached value; otherwise users could get incorrect values when computing multiple forwards then multiple log_abs_det_jacobians. Something like this seems safer:
```py
def __init__(self, ...):
    ...
    self._cache_log_detJ = None, None, None

def _call(self, x):
    y, log_detJ = ...
    self._cache_log_detJ = x, y, log_detJ
    return y

def _inverse(self, y):
    x, log_detJ = ...
    self._cache_log_detJ = x, y, -log_detJ
    return x

def lob_abs_det_jacobian(self, x, y):
    old_x, old_y, log_detJ = self._cache_log_detJ
    if x is not x_old or y is not y_old:
        raise NotImplementedError("out of order .log_abs_det_jacobian() is not implemented")
    return log_detJ
```

Reply:
I don't think this would be a problem. `log_abs_det_jacobian` should only be used internally at https://github.com/pytorch/pytorch/blob/master/torch/distributions/transformed_distribution.py#L112, and `self._cache_log_detJ` is set for each call of `x = transform.inv(y)`, so you can score a given sample multiple times.

For other classes that don't have an analytic inverse, trying to score a sample multiple times with `log_prob()` it will throw an Exception when it tries to find the `x` that corresponds to a given `y`...

So provided you don't use `log_abs_det_jacobian` publicly everything should work as it ought to - what do you think?

Reply:
But  `.log_abs_det_jacobian()` is a public method. I count seven uses in Pyro outside of internal `transforms` code.

Reply:
@stefanwebb it's only a little bit more code to store and check the `x, y` values, and would save some poor grad student a week of frustrated debugging. I'd really prefer to be safe.

Reply:
@fritzo could you please check my slightly different solution? I think it's better to use the cache in the parent transform. If this is okay, I'll make the same change to other transforms where the same problem occurs